### PR TITLE
add `bun publish`

### LINKED ISSUE #1976: Implement `bun publish`
### What is the problem this feature would solve?

Publishing to NPM Registry.

### What is the feature you are proposing to solve the problem?

A `publish` command which makes use of already logged in information through npm's cli

### What alternatives have you considered?

_No response_

### PR DESCRIPTION
**Pull request by dylan-conway**
### What does this PR do?
Implements `bun publish` for publishing workspaces and tarballs

```sh
# To publish the package from the cwd
$ bun publish

# Or provide a path to a tarball
$ bun publish path/to/tarball.tgz
```

closes #1976 

todo:
- [x] improve logging (registry url, included files, response errors, ...)
- [x] read more from `publishConfig`
- [x] fix `--dry-run` for tarball paths
- [x] docs (included in #14327)
- [x] strip build tags from package names

bonus:
- [ ] `--filter`

### How did you verify your code works?
todo:
- [x] publish from tarball path
- [x] publish root workspace and workspace packages
- [x] private package
- [x] restricted access unscoped package
- [x] otp (both web login and classic)
- [x] `--tag`
- [x] tests for `prepublishOnly`, `publish`, and `postpublish` scripts
- [x] also test for pack scripts
- [x] `--ignore-scripts`
- [x] `--dry-run`
- [x] user-agent tests with different ci environment variables (included in #14328)

### CODE REVIEW

**Comment by robobun** [comment]
@dylan-conway, your commit 2a8424e658591b08fda612c6ef817e7a29a08388 is building: [`#4037`](https://buildkite.com/bun/bun/builds/4037)

File: patches/libarchive/archive_write_add_filter_gzip.c.patch [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+ 	}
++	data->os = 3; /* default Unix */
+ 	f->data = data;
+ 	f->open = &archive_compressor_gzip_open;
+ 	f->options = &archive_compressor_gzip_options;
+@@ -166,6 +168,30 @@ archive_compressor_gzip_options(struct archive_write_filter *f, const char *key,
+ 		return (ARCHIVE_OK);
+ 	}
+ 
++	if (strcmp(key, "os") == 0) {
```
Comment (Jarred-Sumner) [inline patches/libarchive/archive_write_add_filter_gzip.c.patch:23]: whats the story here?

File: src/StandaloneModuleGraph.zig [hunk 2]
```diff
@@ -1039,7 +1039,7 @@ pub const StandaloneModuleGraph = struct {
             bun.JSAst.Expr.Data.Store.reset();
             bun.JSAst.Stmt.Data.Store.reset();
         }
-        var json = bun.JSON.ParseJSON(&json_src, &log, arena, false) catch
+        var json = bun.JSON.parseJSON(&json_src, &log, arena, false) catch
```
Comment (Jarred-Sumner) [inline src/StandaloneModuleGraph.zig]: maybe we should make it called `parse`

File: src/bun.js/module_loader.zig [hunk 3]
```diff
@@ -184,9 +184,9 @@ fn dumpSourceStringFailiable(vm: *VirtualMachine, specifier: string, written: []
                 \\  "mappings": "{}"
                 \\}}
             , .{
-                js_printer.formatJSONStringUTF8(std.fs.path.basename(specifier)),
-                js_printer.formatJSONStringUTF8(specifier),
-                js_printer.formatJSONStringUTF8(source_file),
+                bun.fmt.formatJSONStringUTF8(std.fs.path.basename(specifier)),
```
Comment (Jarred-Sumner) [inline src/bun.js/module_loader.zig:187]: jsonStringify?

File: patches/libarchive/archive_write_add_filter_gzip.c.patch (hunk 1 above)
Comment (dylan-conway) [reply patches/libarchive/archive_write_add_filter_gzip.c.patch:23]: gzip has a byte reserved for specifying which operating system it was compressed on. libarchive hardcodes it to 3 (Unix), and the package managers set it to 255 (unknown os).

File: src/StandaloneModuleGraph.zig [hunk 4]
```diff
@@ -1039,7 +1039,7 @@ pub const StandaloneModuleGraph = struct {
             bun.JSAst.Expr.Data.Store.reset();
             bun.JSAst.Stmt.Data.Store.reset();
         }
-        var json = bun.JSON.ParseJSON(&json_src, &log, arena, false) catch
+        var json = bun.JSON.parse(&json_src, &log, arena, false) catch
```
Comment (Jarred-Sumner) [inline src/StandaloneModuleGraph.zig:1042]: nice

**Comment by Eveeifyeve** [comment]
I guess you kinda took my idea to contribute to bun.

**Pull request merged**